lua/nvim-surround/init.lua

```diff
@@ -195,6 +193,34 @@ M.change_surround = function(args)
     local selections = utils.get_nearest_selections(args.del_char, "change")
     local delimiters = args.add_delimiters()
     if selections and delimiters then
+        -- Avoid adding any, and remove any existing whitespace after the
+        -- opening delimiter if only whitespace exists between it and the end
+        -- of the line. Avoid adding or removing leading whitespace before the
+        -- closing delimiter if only whitespace exists between it and the
+        -- beginning of the line.
+
+        -- Column number of end of opening selection
+        local xpos = selections.left.last_pos[2]
+        local space_begin, space_end = buffer.get_line(selections.left.last_pos[1]):find("%s*$")
+
+        if space_begin - 1 <= xpos then -- Whitespace is adjacent to opening delimiter
+            -- Trim trailing whitespace from opening delimiter
+            delimiters[1][#delimiters[1]] = delimiters[1][#delimiters[1]]:gsub("%s+$", "")
```

Comment:
Are these trimmings necessary? Wouldn't they have already been taken into account in the implementation of `get_delimiters`?

Reply:
The simple answer is that they don't get trimmed if I remove either of the `…:gsub(…)` lines. In the `change_surround()` code path, `get_delimiters()` gets called with `line_mode = false`, so no trimming happens there.

Calling `get_delimiters()` with `line_mode = true` isn't right either, since we don't actually want newlines in the delimiters and we would have to call `get_delimiters()` differently depending on whether the matched delimiters are the first/last non-blank, which should happen on a per-delimiter basis which would require an even more complicated call signature along the lines of `get_delimiters(char, line_mode, trim_open_delim, trim_close_delim)`.

I would also prefer not to have to duplicate the trimming logic, but it's not obvious how to refactor the code to do this without a net increase in code volume/complexity.

Reply:
I think pulling out a function that trims delimiters (trailing from left, leading from right) would work, as this logic is used at least twice. Which file to put the function in is a whole other problem (could reside in utils for now, although my goal has been to try and remove utils altogether). Your explanation makes sense though. If you want, you can think more about how you might want to refactor; otherwise the code looks good to me and I can merge first, refactor later :)